### PR TITLE
revised lesson design as suggested by Lesson Development Study Group

### PR DESCRIPTION
**Pull request by psteinb**
This PR is due to my first [homework assignment](https://carpentries-incubator.github.io/study-groups/00-introduction/index.html) from the lesson development study group (LDSG). For this, I revised the design document and adapted the structure suggested by the LDSG.

Some observations:
- I was completely unclear if this material aims to teach deep learning with images or with tables. Both is possible, but I'd feel more comfortable to do image based deep learning. Would love to hear feedback on this?
- I added 3 learner profiles for now, that give away a bit that this material relies on images mostly. 
- it wasn't too clear to me what people intended with "Troubleshoot the learning process" as a learning objective
- I added a jekyll dependencies to allow graphviz plots being included in each page

I'd appreciate feedback until end of this week so that I can incorporate/adapt to it until the next LDSG session.

### CODE REVIEW

**Comment by dafnevk** [comment]
Hi Peter, thanks for the pull request, I will have a look at it tomorrow so I can provide you with feedback this week.

File: _extras/design.md [hunk 1]
```diff
@@ -4,35 +4,159 @@ title: Lesson Design
...
+- critique the results of a partners network
+
+### Re-use existing network architectures with and without pre-trained weights
+
+Re-use of architectures is common in deep learning. Especially when using pre-trained weights (transfer-learning) it can also be very powerful.
+
+After this module, learners can ...
+
+- describe what transfer learning stands for
+- explain the expected relation of data sets (one used for the network originally and one for the network at hand)
```
Comment (dafnevk) [inline _extras/design.md:133]: maybe word it slightly different:
explain in what situations transfer learning is beneficial.

**Review by dafnevk — COMMENTED** [review]
Thanks Peter for the effort, this really helps moving the lesson design forward.

About image / tabular data: Our idea was that this lesson can be a quite generic introduction, where we first apply neural networks on tabular data to explain the basic concepts, and then on image data to explain convolutional layers.
I think this can be useful to a wide audience of researchers, even if they don't have image data, it is often useful to learn about concepts such as convolutional layers. The only data type we chose to explicitly leave out is text data, because that requires some specific techniques (word embeddings and contextualized word embeddings) that would take too much effort to explain to a heterogeneous learners group.

I realize we didn't share the learner profiles that @bpmweel wrote for this lesson, but at least that gave you a chance to come up with fresh ones for the study group. I left a few comments about the ones you propose.

The detailed definition of the learning goals is very useful, and makes explicit in how much depth this lesson should go. I left a few comments on items that I found unclear or disagreed with.

Before merging this pull request, maybe @annefou @bpmweel and @florian-huber (or a subset of these people) can have the opportunity to have a look, since we wrote down the initial lesson design with this group of people.

**Comment by dafnevk** [comment]
I've been copy-pasting things all over the place, and now it seems like my comments on the learner profiles are lost. I'll try to rephrase them again:
- It would be good to also include a profile of a researcher with non-image data
- The researchers in all three profiles already have collected and annotated data, ready for deep learning. I also expect scenarios where researchers have not yet collected data (and can be inspired through this lesson what is possible with deep learning), or they have data but it is not annotated or too small.
- You can make more explicit what the challenges are that the researchers currently face, and how this course on deep learning could help them. Eg.: are there current methods not accurate enough? Or not fast enough? Did their supervisor tell them to use Deep learning? Does everyone in the field already do it?  Do they want deep learning on their resume? etc.

**Comment by dafnevk** [comment]
When I build the site locally (with `bundle update` and `make serve`) the graphviz plot does not render, and I get the error `Jekyll Diagrams: Command Not Found: dot`. Am I missing a step in the installation?

File: _extras/design.md [hunk 2]
```diff
@@ -4,35 +4,159 @@ title: Lesson Design
...
+
+The following offers more details to each learning objective based on Bloom's Taxonomy. For hints on how to use this approach, see [lesson 15 of the instructor training](https://carpentries.github.io/instructor-training/15-lesson-study/index.html)
+
+### Prepare input data for use for deep learning 
+
+This includes cleaning data, filling missing values, normalizing, and transforming categorical columns into dummy encoding.
+
+After this module, learners can ...
+
+- define a checklist how to analyze the raw data considering an application of ML
```
[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md:64]: With this, I wanted to say: Learners can take pen and paper and outline a checklist what aspects of the data they would check BEFORE data is handed to a learning algorithm of any kind. For example, I'd check scales of variables, histograms of important features, ...

File: _extras/design.md [hunk 3]
```diff
@@ -4,35 +4,159 @@ title: Lesson Design
...
+- describe criteria by which to judge good or bad data, e.g. how a column's values should be distributed
+- execute a min-max normalization on floating point data
+- sketch how to insert missing timestamps or literal values (i.e. factors or non-numeric entries)
+- implement a transformation of categorical values into a numerical encoding (`int8`)
+- argue for or against strategies to to normalize data
+- formulate techniques beyond transformations to clean data (exploit correlations, create summary statistics, ...)
```
[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md:70]: OK, I paraphrased this from the minimal learning goals that were there before. I am fine with leaving things out. When thinking about deep learning (or CNNs) for image data, there are mostly 2 operations that I consider essential: 
- min-max scaling of pixel intensities
- cropping/padding images to a consistent size

File: _extras/design.md [hunk 4]
```diff
@@ -4,35 +4,159 @@ title: Lesson Design
...
+
+### Design and train a Deep Neural Network
+
+This includes knowledge of when to different types of layers
+
+After this module, learners can ...
+
+- list/repeat the three ingrediants to a feed forward network: input, hidden layers, output
+- classify parts of a feed forward network
```
[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md:79]: When I present a network architecture of a feed forward network to my learner, say like the output of `keras.model.summary()`, she can tell me what the acronyms stand for and what they do.

[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md:79]: I updated this with 67d4065

File: _extras/design.md [hunk 5]
```diff
@@ -4,35 +4,159 @@ title: Lesson Design
...
+- describe a fully connected (dense) layer
+- describe a convolutional layer
+- describe a max pooling layer
+- describe an activation function
+- describe a softmax layer
+- argue against abundant use of the sigmoid function (exploding/vanishing gradients)
+- select the output data shape from a convolutional layer
```
[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md]: Well spotted. I meant calculate. Fixed with bb37279

File: _extras/design.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md:70]: Let me know which items you think can stay or should be added.

File: _extras/design.md [hunk 6]
```diff
@@ -4,35 +4,159 @@ title: Lesson Design
...
+- interpret errors with convolutional layers
+- execute a 3 layer network on the MNIST data (or similar)
+- differentiate a dense layer and a convolutional layer
+- experiment with values of dense layer and a convolutional layer
+- select a layer type depending on the input data
+- develop a 5 layer network that comprises both layer types
+
+### Troubleshoot the learning process
+
+Often when designing neural networks training will not automatically work very well. This requires setting the parameters of the training algorithm correctly, modifying the design of the network or changing the data pre-processing
```
[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md]: fixed with f5f5040

File: _extras/design.md (hunk 1 above)
Comment (psteinb) [reply _extras/design.md]: fixed with d5fe467

**Comment by psteinb** [comment]
For the graphviz stuff, you need `dot` installed. Let me know if that is counterproductive or if there is a jekyll plugin that could do better.

**Comment by psteinb** [comment]
> I've been copy-pasting things all over the place, and now it seems like my comments on the learner profiles are lost. I'll try to rephrase them again:
> 
>     * It would be good to also include a profile of a researcher with non-image data
> 

fixed with 163d11c :heavy_check_mark: 

>     * The researchers in all three profiles already have collected and annotated data, ready for deep learning. I also expect scenarios where researchers have not yet collected data (and can be inspired through this lesson what is possible with deep learning), or they have data but it is not annotated or too small.
> 

good point - tried to relax the language a bit with 00ad34b and hope this would meet your points better. Feel free to suggest another learner profile if you feel that an important part is missing.

>     * You can make more explicit what the challenges are that the researchers currently face, and how this course on deep learning could help them. Eg.: are there current methods not accurate enough? Or not fast enough? Did their supervisor tell them to use Deep learning? Does everyone in the field already do it?  Do they want deep learning on their resume? etc.

I see your point. Done with 9b2de00. if you think this is not enough, just drop concrete text into this PR.

File: _extras/design.md [hunk 7]
```diff
@@ -4,35 +4,160 @@ title: Lesson Design
...
+- They don’t know if it’s the right tool for their situations. 
+- They have no idea what it takes to actually do deep learning.
+- Want to quickly have some useful skills for their own data.
+
+#### Required Pre-Knowledge
 
-- **Basic Machine Learning Knowledge** – Data Cleaning, Train & Test Split, Overfitting & Underfitting, Metrics (Accuracy, Recall, etc.),
 - **Python** – Previous Experience programming in python is required (Refer to Python Data Carpentry Lesson )
 - **Pandas** – Knowledge of the Pandas python package
+- **Basic Machine Learning Knowledge** – Data Cleaning, Train & Test Split, Overfitting & Underfitting, Metrics (Accuracy, Recall, etc.),
```
Comment (florian-huber) [inline _extras/design.md:42]:
```suggestion
- **Python** – Previous programming experience in Python is required (Refer to Python Data Carpentry Lesson)
- **Pandas** – Knowledge of the Pandas Python package
- **Basic Machine Learning Knowledge** – Data cleaning, train & test split, overfitting & underfitting, metrics (accuracy, recall, etc.),
```

Comment (florian-huber) [reply _extras/design.md:42]: only nitty-gritty stuff, trying to make it more consistent (feel free to change/ignore).

File: _extras/design.md [hunk 8]
```diff
@@ -4,35 +4,160 @@ title: Lesson Design
...
 
-- **Prepare input data for use for deep learning** – This includes cleaning data, filling missing values, normalizing, and transforming categorical columns into dummy encoding.
+> ## Overview
+> After following this lesson, learners will be able to:
+>  
+> - Prepare input data for use for deep learning 
+> - Design and train a Deep Neural Network
+> - Troubleshoot the learning process
+> - Measure the performance of the network
+> - Visualizing Data and Results
```
Comment (florian-huber) [inline _extras/design.md:53]:
```suggestion
> - Visualizing data and results
```

File: _extras/design.md [hunk 9]
```diff
@@ -4,35 +4,160 @@ title: Lesson Design
...
+- formulate techniques beyond transformations to clean data (exploit correlations, create summary statistics, ...)
+
+### Design and train a Deep Neural Network
+
+This includes knowledge of when to different types of layers
+
+After this module, learners can ...
+
+- list/repeat the three ingrediants to a feed forward network: input, hidden layers, output
+- classify/categorize parts of a feed forward network when presented a network architecture (as from `keras.model.summary()`
```
Comment (florian-huber) [inline _extras/design.md:82]:
```suggestion
- list/repeat the three ingredients to a feed forward network: input, hidden layers, output
- classify/categorize parts of a feed forward network when presented a network architecture (as from `keras.model.summary()`)
```

File: _extras/design.md [hunk 10]
```diff
@@ -4,35 +4,160 @@ title: Lesson Design
...
+After this module, learners can ...
+
+- define precision and recall/accuracy for a classification task
+- state that cross-validation is used in Deep Learning too
+- describe how to split a dataset into training/test/validation set
+- describe how Drop-Out Layers work
+- execute a plot to draw the loss per epoch for training and test set
+- compare values of precision and recall
+- differentiate a overfitting network from a well-behaved network
+- detect a underfitting network
```
Comment (florian-huber) [inline _extras/design.md:110]:
```suggestion
- detect when a network is underfitting or overfitting
```

File: _extras/design.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (florian-huber) [reply _extras/design.md:73]: Could we reword this to make clearer what is meant, e.g. 'techniques to prepare (cleaned) data for training a deep learning network` or similar?

Because I agree that there are some preparation steps (as the 2 mentioned by Peter above) that are rather specific for DL. The re-scaling is also relevant for any feature, not only pixel intensities.

**Comment by florian-huber** [comment]
Really nice developments, thanks a lot @psteinb (and @dafnevk)!
I added a few tiny comments/edits. 
- Apart from that Dafne mentioned all main points I had, in particular that I would also prefer to make this not purely image-data centered. My reason for that is that DL on image-data is already covered well by many tutorials and quiet established (including clear-ish network design principles). But it gets much harder to get started if people have different data to work with. That said, I liked the list of learning goals quiet a bit!
- On the learning profiles: I think it would be great if @bpmweel could have a look at those since he was working on that as well.

File: _extras/design.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md:67]: @dafnevk let me know if any rewording is required. Otherwise, I'll resolve this comment.

File: _extras/design.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md:73]: I see multiple ways of going forward with this. I'd love to remain within Bloom's taxonomy though as this helped me tremendously. I don't know if @dafnevk suggested to remove specific points from the list above. I assume your comment focused on `-  formulate techniques beyond transformations ...`, right?

How about this then:
```suggestion
- formulate techniques to prepare data for training a deep learning network (homogenous use of encodings, summary statistics to check if they appear consistent, min/max values,  ...)
```

[the remark this answers is not in the document]
Comment (dafnevk) [reply _extras/design.md:73]: Much better and specific, although now it overlaps a bit with the learning goals above it. And I don't fully understand what you mean with 'to check if they appear consistent'.
Would it make sense to put it as the first learning goal in this category:
```suggestion
- formulate techniques to prepare (clean) data for training a deep learning network
```
And then specifying the techiniques further in the separate learning goals as you wrote them down already (execute min-max normalization, etc).

File: _extras/design.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dafnevk) [reply _extras/design.md:67]: Ok that is useful indeed, maybe reword it as follows:
```suggestion
- define a checklist for data analysis steps before applying Deep Learning to the data
```

File: _extras/design.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (psteinb) [reply _extras/design.md]: sounds good.

**Comment by psteinb** [comment]
:weary: my local git cannot merge anymore. Something is at odds. If anyone could resolve the conflict, please help!

**Comment by dafnevk** [comment]
> 😩 my local git cannot merge anymore. Something is at odds. If anyone could resolve the conflict, please help!

What is the merge conflict that you get?
The pull request mentions that the merge problem originates from the .travis.yml file, that was removed in the gh-pages branch but you edited it for this pull request. Does it solve your conflict if you remove the .travis.yml file? We don't need it anymore since we switched to github pages.

**Comment by psteinb** [comment]
I give that a shot. I think the basic problem stems from the fact I commit changes in the github web interface and work on the same file locally. If I try to merge from my fork then, `git` tells me that I can't do that.

**Comment by psteinb** [comment]
```
$ git pull origin lesson-development-study-group-design
From github.com:psteinb/deep-learning-intro
 * branch            lesson-development-study-group-design -> FETCH_HEAD
fatal: Not possible to fast-forward, aborting.
```

**Comment by psteinb** [comment]
Ah, it was my mistake ... my config has `git config --global pull.ff` set to `True`. Hence, a simple `git pull --no-ff` solves the issue. Feel free to merge this once possible.

**Comment by dafnevk** [comment]
Thanks Peter, only one thing before I merge the PR: even though the GH actions build does not fail, it does give the error `Jekyll Diagrams: Command Not Found: dot`. So I think installation of graphviz needs to be added to the GH actions config file. Also, maybe this can be added to the README as installation instructions for running the lesson website locally.

**Comment by tobyhodges** [comment]
dropping by to add an apology here: I included the GraphViz concept mapping in the template for the lesson design notes but did not think about compatibility between that and the lesson template. If community members are finding this useful as a tool for lesson design, we might be able to incorporate it into an upcoming version of the lesson template. For now, if you can find a solution that works for your lesson repository, that would be great.

**Comment by psteinb** [comment]
@tobyhodges I find it useful as the concept maps are useful to some people. Can you point me at the container image that the actions for this lessons use? I am not sure how to bring in graphviz for the time being.

@dafnevk let me see what I can do.

**Comment by tobyhodges** [comment]
@psteinb the Actions workflow is defined in https://github.com/carpentries-incubator/deep-learning-intro/blob/gh-pages/.github/workflows/website.yml - I _think_ you can add GraphViz as a jekyll plugin there, perhaps during the `gem install` phase?

**Comment by psteinb** [comment]
I did that locally already, but the gem doesn't bring the `dot` app with it in any fashion - at least on my fc33 box. So I'd
need a way to apt-install or dnf-install packages if possible.

Ahoi! I just saw that the actions script is based on ubuntu-20.04. Ok, I didn't see that before for some reason. Then I guess, I
am good to go.

**Comment by dafnevk** [comment]
Thanks Peter, checks are passing so I'll merge!

**Comment by dafnevk** [comment]
Hm, even though the GH actions checks were green for this PR, the website won't build. (See [here](https://github.com/carpentries-incubator/deep-learning-intro/runs/2087381037) ). It seems that indeed graphviz is not a supported jekyll plugin. 
@tobyhodges do you know if this can be fixed? (are there any other lesson websites that use the graphviz plugin?)

**Comment by psteinb** [comment]
it builds fine locally. I don't understand this to be fair, the github actions log appears to be sound, but the website is not there, right?

**Comment by tobyhodges** [comment]
As I understand it, the issue here is that the GitHub Actions workflow builds the lesson pages in a container, recreating the environment used by GitHub Pages, to check that everything is fine and give some feedback if there is a problem. But the pages built there are not the pages served by GitHub Pages; that build is done elsewhere. So, you specified that graphviz should be included in the Actions workflow, and that is working fine. But the _actual_ GitHub Pages run still does not include graphviz or the `jekyll-diagrams` plugin (and cannot be configured to do so) so your pages do not build 😞 pinging @fmichonneau to confirm or correct my theory.

It increasingly looks like it was a bad idea for me to include GraphViz in the CodiMD template for these lesson design notes - sorry for the trouble!

**Comment by fmichonneau** [comment]
@tobyhodges yes, you are correct. The pages are built by GitHub itself, we recreate the same environment that they use to check that there won't be any issues during the build process (and see what the issue is if there is a problem).

For plugins, we are limited to what comes with the plugins that come with the `github-pages` gem: https://pages.github.com/versions/ 

Currently, the pages in the `_extra` folder don't support Rmarkdown rendering but if the graph was in an episode, then we could render it by using the `DiagrammeR` [package](http://rich-iannone.github.io/DiagrammeR/graphviz_and_mermaid.html).

**Comment by tobyhodges** [comment]
Thank you @fmichonneau 👍 

For now, @psteinb I recommend that you revert the changes to the Actions workflow and comment out this part of the Design Notes page, ready to re-introduce it if and when a future version of the lesson template supports this kind of diagram.

**Comment by psteinb** [comment]
I see @dafnevk was faster than me.
